File: android-sdk/src/main/java/com/maze/sdk/Maze.kt
package com.maze.sdk

import android.app.Activity
import android.app.Application
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.ByteArrayOutputStream
import java.io.File
import java.lang.ref.WeakReference
import java.time.Instant
import java.util.UUID
import java.util.concurrent.ConcurrentLinkedQueue

private val recommendedBlockedCaptureScreens: Set<String> = setOf(
    "login",
    "signup",
    "otp_verification",
    "password_reset",
    "payment",
    "kyc_id_upload",
)

data class MazeEvent(
    val event_id: String,
    val session_id: String,
    val device_id: String,
    val occurred_at: String,
    val event: String,
    val screen: String?,
    val element_id: String?,
    val x: Float?,
    val y: Float?,
    val screen_width: Float?,
    val screen_height: Float?,
    val app_version: String?,
    val screenshot_id: String?,
    val metadata: Map<String, String>
)

object MazeSession {
    private var sessionId: String = UUID.randomUUID().toString()

    fun currentSessionId(): String = sessionId

    fun reset() {
        sessionId = UUID.randomUUID().toString()
    }
}

private class NetworkClient(
    private val endpoint: String,
    private val screenshotEndpoint: String,
    private val apiKey: String,
    private val client: OkHttpClient = OkHttpClient()
) {
    private fun escapeJson(raw: String): String {
        return raw.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n")
    }

    fun send(events: List<MazeEvent>) {
        val jsonEvents = events.joinToString(separator = ",") { event ->
            """
            {
              "event_id":"${escapeJson(event.event_id)}",
              "session_id":"${escapeJson(event.session_id)}",
              "device_id":"${escapeJson(event.device_id)}",
              "occurred_at":"${escapeJson(event.occurred_at)}",
              "event":"${escapeJson(event.event)}",
              "screen":${event.screen?.let { "\"${escapeJson(it)}\"" } ?: "null"},
              "element_id":${event.element_id?.let { "\"${escapeJson(it)}\"" } ?: "null"},
              "x":${event.x?.toString() ?: "null"},
              "y":${event.y?.toString() ?: "null"},
              "screen_width":${event.screen_width?.toString() ?: "null"},
              "screen_height":${event.screen_height?.toString() ?: "null"},
              "app_version":${event.app_version?.let { "\"${escapeJson(it)}\"" } ?: "null"},
              "screenshot_id":${event.screenshot_id?.let { "\"${escapeJson(it)}\"" } ?: "null"},
              "metadata":${event.metadata.entries.joinToString(prefix = "{", postfix = "}") { "\"${escapeJson(it.key)}\":\"${escapeJson(it.value)}\"" }}
            }
            """.trimIndent()
        }

        val body = """{"events":[$jsonEvents]}""".toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url(endpoint)
            .addHeader("X-API-Key", apiKey)
            .post(body)
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                error("Failed to upload events: ${response.code}")
            }
        }
    }

    fun uploadScreenshot(
        imageBytes: ByteArray,
        screen: String?,
        sessionId: String,
        width: Int,
        height: Int
    ): String {
        val tempFile = File.createTempFile("maze_capture_", ".jpg")
        tempFile.writeBytes(imageBytes)
        try {
            val body = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("session_id", sessionId)
                .addFormDataPart("screen", screen ?: "")
                .addFormDataPart("width", width.toString())
                .addFormDataPart("height", height.toString())
                .addFormDataPart(
                    "screenshot",
                    "capture.jpg",
                    tempFile.asRequestBody("image/jpeg".toMediaType())
                )
                .build()
            val request = Request.Builder()
                .url(screenshotEndpoint)
                .addHeader("X-API-Key", apiKey)
                .post(body)
                .build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    error("Failed to upload screenshot: ${response.code}")
                }
                val payload = response.body?.string() ?: return ""
                val match = Regex("\"screenshot_id\"\\s*:\\s*\"([^\"]+)\"").find(payload)
                return match?.groupValues?.getOrNull(1) ?: ""
            }
        } finally {
            tempFile.delete()
        }
    }
}

private class EventQueue(
    private val networkClient: NetworkClient,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
) {
    private val buffer = ConcurrentLinkedQueue<MazeEvent>()
    private var scheduledFlush: Job? = null

    fun enqueue(event: MazeEvent) {
        buffer.add(event)
        if (buffer.size >= 20) {
            flush()
            return
        }
        if (scheduledFlush == null) {
            scheduledFlush = scope.launch {
                delay(5_000)
                flush()
            }
        }
    }

    fun flush() {
        val events = mutableListOf<MazeEvent>()
        while (true) {
            val event = buffer.poll() ?: break
            events.add(event)
        }
        scheduledFlush?.cancel()
        scheduledFlush = null

        if (events.isEmpty()) {
            return
        }

        scope.launch {
            runCatching {
                networkClient.send(events)
            }.onFailure {
                events.forEach(buffer::add)
                delay(2_000)
                flush()
            }
        }
    }
}

data class MazeConfig(
    val apiKey: String,
    val deviceId: String,
    val endpoint: String = "http://10.0.2.2:8000/events",
    val appVersion: String? = null,
    val sessionCaptureEnabled: Boolean = false,
    val screenshotQuality: Int = 72,
    val screenshotMaxDimension: Int = 1280,
    val screenshotEndpoint: String = endpoint.replace("/events", "/screenshots"),
    val application: Application? = null,
    val captureAllowedScreens: Set<String>? = null,
    val captureBlockedScreens: Set<String> = recommendedBlockedCaptureScreens,
    val captureStatusListener: ((Boolean) -> Unit)? = null,
    val captureEvaluator: ((String?) -> Boolean)? = null,
)

object ActivityTracker : Application.ActivityLifecycleCallbacks {
    private var currentRef: WeakReference<Activity>? = null
    private var attached = false

    fun attach(app: Application) {
        if (attached) {
            return
        }
        app.registerActivityLifecycleCallbacks(this)
        attached = true
    }

    fun current(): Activity? = currentRef?.get()

    override fun onActivityResumed(activity: Activity) {
        currentRef = WeakReference(activity)
    }

    override fun onActivityStarted(activity: Activity) {}
    override fun onActivityPaused(activity: Activity) {}
    override fun onActivityStopped(activity: Activity) {}
    override fun onActivitySaveInstanceState(activity: Activity, outState: android.os.Bundle) {}
    override fun onActivityCreated(activity: Activity, savedInstanceState: android.os.Bundle?) {}
    override fun onActivityDestroyed(activity: Activity) {}
}

object Maze {
    private var currentScreen: String? = null
    private var config: MazeConfig? = null
    private var queue: EventQueue? = null
    private var networkClient: NetworkClient? = null
    private val workerScope = CoroutineScope(Dispatchers.IO)

    @Volatile private var sessionCaptureEnabled: Boolean = false
    @Volatile private var captureAllowedScreens: Set<String>? = null
    @Volatile private var captureBlockedScreens: Set<String> = emptySet()
    @Volatile private var captureStatusListener: ((Boolean) -> Unit)? = null
    @Volatile private var captureEvaluator: ((String?) -> Boolean)? = null

    fun configure(config: MazeConfig) {
        validateEndpoint(config.endpoint)
        this.config = config
        sessionCaptureEnabled = config.sessionCaptureEnabled
        captureAllowedScreens = config.captureAllowedScreens
        captureBlockedScreens = config.captureBlockedScreens
        captureStatusListener = config.captureStatusListener
        captureEvaluator = config.captureEvaluator
        if (config.application != null) {
            ActivityTracker.attach(config.application)
        }
        networkClient = NetworkClient(config.endpoint, config.screenshotEndpoint, config.apiKey)
        queue = EventQueue(requireNotNull(networkClient))
        captureStatusListener?.invoke(sessionCaptureEnabled)
    }

    fun setSessionCaptureEnabled(enabled: Boolean) {
        sessionCaptureEnabled = enabled
        captureStatusListener?.invoke(enabled)
    }

    fun setCaptureAllowedScreens(screens: Set<String>?) {
        captureAllowedScreens = screens
    }

    fun setCaptureBlockedScreens(screens: Set<String>) {
        captureBlockedScreens = screens
    }

    fun setScreenCaptureEnabled(enabled: Boolean, screen: String) {
        captureBlockedScreens = if (enabled) captureBlockedScreens - screen else captureBlockedScreens + screen
    }

    fun screen(name: String) {
        currentScreen = name
        val activeConfig = requireNotNull(config) { "Call Maze.configure(MazeConfig) before tracking events." }
        val activeQueue = requireNotNull(queue) { "Tracker queue is not initialized." }
        val client = requireNotNull(networkClient) { "Network client is not initialized." }
        workerScope.launch {
            val screenshotId = if (shouldCapture(name)) {
                runCatching { captureAndUploadScreenshot(activeConfig, client, name) }.getOrNull()
            } else {
                null
            }
            enqueueEvent(activeConfig, activeQueue, "screen_view", name, null, emptyMap(), null, null, screenshotId)
        }
    }

    fun track(
        event: String,
        screen: String? = null,
        elementId: String? = null,
        metadata: Map<String, String> = emptyMap(),
        x: Float? = null,
        y: Float? = null
    ) {
        val activeConfig = requireNotNull(config) { "Call Maze.configure(MazeConfig) before tracking events." }
        val activeQueue = requireNotNull(queue) { "Tracker queue is not initialized." }
        workerScope.launch {
            enqueueEvent(activeConfig, activeQueue, event, screen ?: currentScreen, elementId, metadata, x, y, null)
        }
    }

    private fun shouldCapture(screen: String?): Boolean {
        if (!sessionCaptureEnabled) return false
        if (captureEvaluator?.invoke(screen) == false) return false
        if (screen != null && captureBlockedScreens.contains(screen)) return false
        val allowed = captureAllowedScreens
        return allowed == null || (screen != null && allowed.contains(screen))
    }

    private suspend fun enqueueEvent(
        activeConfig: MazeConfig,
        activeQueue: EventQueue,
        event: String,
        screen: String?,
        elementId: String?,
        metadata: Map<String, String>,
        x: Float?,
        y: Float?,
        screenshotId: String?
    ) {
        val safeMetadata = metadata.mapValues { (_, value) -> if (value.length > 24) "***" else value }
        val displayMetrics = Resources.getSystem().displayMetrics
        val normalizedX = x?.let { (it / displayMetrics.widthPixels).coerceIn(0f, 1f) }
        val normalizedY = y?.let { (it / displayMetrics.heightPixels).coerceIn(0f, 1f) }

        activeQueue.enqueue(
            MazeEvent(
                event_id = UUID.randomUUID().toString(),
                session_id = MazeSession.currentSessionId(),
                device_id = activeConfig.deviceId,
                occurred_at = Instant.now().toString(),
                event = event,
                screen = screen,
                element_id = elementId,
                x = normalizedX,
                y = normalizedY,
                screen_width = displayMetrics.widthPixels.toFloat(),
                screen_height = displayMetrics.heightPixels.toFloat(),
                app_version = activeConfig.appVersion,
                screenshot_id = screenshotId,
                metadata = safeMetadata
            )
        )
    }

    private suspend fun captureAndUploadScreenshot(
        activeConfig: MazeConfig,
        client: NetworkClient,
        screenName: String
    ): String? {
        val capture = withContext(Dispatchers.Main) {
            val activity = ActivityTracker.current() ?: return@withContext null
            val root = activity.window?.decorView?.rootView ?: return@withContext null
            if (root.width <= 0 || root.height <= 0) return@withContext null
            val bitmap = Bitmap.createBitmap(root.width, root.height, Bitmap.Config.RGB_565)
            val canvas = Canvas(bitmap)
            root.draw(canvas)
            val resized = resizeBitmap(bitmap, activeConfig.screenshotMaxDimension)
            val output = ByteArrayOutputStream()
            resized.compress(Bitmap.CompressFormat.JPEG, activeConfig.screenshotQuality.coerceIn(35, 90), output)
            val bytes = output.toByteArray()
            output.close()
            val width = resized.width
            val height = resized.height
            if (resized != bitmap) resized.recycle()
            bitmap.recycle()
            Triple(bytes, width, height)
        } ?: return null

        val screenshotId = client.uploadScreenshot(
            imageBytes = capture.first,
            screen = screenName,
            sessionId = MazeSession.currentSessionId(),
            width = capture.second,
            height = capture.third
        )
        return screenshotId.ifBlank { null }
    }

    private fun resizeBitmap(source: Bitmap, maxDimension: Int): Bitmap {
        val maxSide = maxOf(source.width, source.height)
        if (maxSide <= maxDimension || maxDimension <= 0) return source
        val scale = maxDimension.toFloat() / maxSide.toFloat()
        val matrix = Matrix().apply { postScale(scale, scale) }
        return Bitmap.createBitmap(source, 0, 0, source.width, source.height, matrix, true)
    }

    private fun validateEndpoint(endpoint: String) {
        val isSecure = endpoint.startsWith("https://")
        val isLocal = endpoint.startsWith("http://127.0.0.1") || endpoint.startsWith("http://10.0.2.2") || endpoint.startsWith("http://localhost")
        check(isSecure || isLocal) { "Maze requires HTTPS for production endpoints." }
    }
}

File: ios-sdk/Sources/Maze/Maze.swift
import Foundation
import UIKit

public actor MazeSession {
    public static let shared = MazeSession()
    private var sessionId: String = UUID().uuidString

    public func currentSessionId() -> String { sessionId }

    public func reset() {
        sessionId = UUID().uuidString
    }
}

public struct MazeConfig: Sendable {
    public static let recommendedBlockedScreens = [
        "login",
        "signup",
        "otp_verification",
        "password_reset",
        "payment",
        "kyc_id_upload",
    ]

    public let apiKey: String
    public let deviceId: String
    public let endpoint: URL
    public let screenshotEndpoint: URL
    public let appVersion: String?
    public let sessionCaptureEnabled: Bool
    public let screenshotQuality: CGFloat
    public let screenshotMaxDimension: CGFloat
    public let captureAllowedScreens: [String]?
    public let captureBlockedScreens: [String]
    public let captureStatusHandler: (@Sendable (Bool) -> Void)?
    public let captureEvaluator: (@Sendable (String?) -> Bool)?

    public init(
        apiKey: String,
        deviceId: String,
        endpoint: URL = URL(string: "http://127.0.0.1:8000/events")!,
        appVersion: String? = nil,
        sessionCaptureEnabled: Bool = false,
        screenshotQuality: CGFloat = 0.72,
        screenshotMaxDimension: CGFloat = 1280,
        captureAllowedScreens: [String]? = nil,
        captureBlockedScreens: [String] = MazeConfig.recommendedBlockedScreens,
        captureStatusHandler: (@Sendable (Bool) -> Void)? = nil,
        captureEvaluator: (@Sendable (String?) -> Bool)? = nil
    ) {
        self.apiKey = apiKey
        self.deviceId = deviceId
        self.endpoint = endpoint
        self.screenshotEndpoint = endpoint.deletingLastPathComponent().appendingPathComponent("screenshots")
        self.appVersion = appVersion
        self.sessionCaptureEnabled = sessionCaptureEnabled
        self.screenshotQuality = screenshotQuality
        self.screenshotMaxDimension = screenshotMaxDimension
        self.captureAllowedScreens = captureAllowedScreens
        self.captureBlockedScreens = captureBlockedScreens
        self.captureStatusHandler = captureStatusHandler
        self.captureEvaluator = captureEvaluator
    }
}

public struct MazeEvent: Codable {
    let eventId: String
    let sessionId: String
    let deviceId: String
    let occurredAt: String
    let event: String
    let screen: String?
    let elementId: String?
    let x: Double?
    let y: Double?
    let screenWidth: Double?
    let screenHeight: Double?
    let appVersion: String?
    let screenshotId: String?
    let metadata: [String: String]

    enum CodingKeys: String, CodingKey {
        case eventId = "event_id"
        case sessionId = "session_id"
        case deviceId = "device_id"
        case occurredAt = "occurred_at"
        case event
        case screen
        case elementId = "element_id"
        case x
        case y
        case screenWidth = "screen_width"
        case screenHeight = "screen_height"
        case appVersion = "app_version"
        case screenshotId = "screenshot_id"
        case metadata
    }
}

actor NetworkClient {
    private let config: MazeConfig
    private let session: URLSession

    init(config: MazeConfig, session: URLSession = .shared) {
        self.config = config
        self.session = session
    }

    func send(events: [MazeEvent]) async throws {
        var request = URLRequest(url: config.endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(config.apiKey, forHTTPHeaderField: "X-API-Key")
        request.httpBody = try JSONEncoder().encode(["events": events])
        _ = try await session.data(for: request)
    }

    func uploadScreenshot(
        jpegData: Data,
        screen: String?,
        sessionId: String,
        width: Int?,
        height: Int?
    ) async throws -> String {
        let boundary = "Boundary-\(UUID().uuidString)"
        var request = URLRequest(url: config.screenshotEndpoint)
        request.httpMethod = "POST"
        request.setValue(config.apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        func appendField(_ name: String, _ value: String) {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(value)\r\n".data(using: .utf8)!)
        }

        appendField("session_id", sessionId)
        if let screen { appendField("screen", screen) }
        if let width { appendField("width", String(width)) }
        if let height { appendField("height", String(height)) }

        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"screenshot\"; filename=\"capture.jpg\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
        body.append(jpegData)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        request.httpBody = body
        let (data, _) = try await session.data(for: request)
        let decoded = try JSONDecoder().decode([String: String].self, from: data)
        if let screenshotId = decoded["screenshot_id"] {
            return screenshotId
        }
        throw NSError(domain: "Maze", code: 1, userInfo: [NSLocalizedDescriptionKey: "screenshot_id missing in response"])
    }
}

actor EventQueue {
    private let flushIntervalNanoseconds: UInt64 = 5_000_000_000
    private let maxBatchSize = 20
    private let client: NetworkClient
    private var buffer: [MazeEvent] = []
    private var flushTask: Task<Void, Never>?

    init(client: NetworkClient) {
        self.client = client
    }

    func enqueue(_ event: MazeEvent) {
        buffer.append(event)
        scheduleFlushIfNeeded()
        if buffer.count >= maxBatchSize {
            Task { await flush() }
        }
    }

    private func scheduleFlushIfNeeded() {
        guard flushTask == nil else { return }
        flushTask = Task {
            try? await Task.sleep(nanoseconds: flushIntervalNanoseconds)
            await flush()
        }
    }

    func flush() async {
        guard !buffer.isEmpty else {
            flushTask = nil
            return
        }

        let payload = buffer
        buffer.removeAll()
        flushTask = nil

        do {
            try await client.send(events: payload)
        } catch {
            buffer.insert(contentsOf: payload, at: 0)
            flushTask = Task {
                try? await Task.sleep(nanoseconds: 2_000_000_000)
                await flush()
            }
        }
    }

    func uploadScreenshot(
        jpegData: Data,
        screen: String?,
        sessionId: String,
        width: Int?,
        height: Int?
    ) async throws -> String {
        try await client.uploadScreenshot(
            jpegData: jpegData,
            screen: screen,
            sessionId: sessionId,
            width: width,
            height: height
        )
    }
}

private final class MazeClient: @unchecked Sendable {
    static let shared = MazeClient()

    private let isoFormatter = ISO8601DateFormatter()
    private let lock = NSLock()
    private var currentScreen: String?
    private var config: MazeConfig?
    private var queue: EventQueue?
    private var sessionCaptureEnabled = false
    private var captureAllowedScreens: Set<String>?
    private var captureBlockedScreens = Set<String>()
    private var captureEvaluator: (@Sendable (String?) -> Bool)?
    private var captureStatusHandler: (@Sendable (Bool) -> Void)?

    private init() {}

    func configure(_ config: MazeConfig) {
        validateEndpoint(config.endpoint)
        lock.lock()
        self.config = config
        queue = EventQueue(client: NetworkClient(config: config))
        currentScreen = nil
        sessionCaptureEnabled = config.sessionCaptureEnabled
        captureAllowedScreens = config.captureAllowedScreens.map { Set($0) }
        captureBlockedScreens = Set(config.captureBlockedScreens)
        captureEvaluator = config.captureEvaluator
        captureStatusHandler = config.captureStatusHandler
        lock.unlock()
        config.captureStatusHandler?(config.sessionCaptureEnabled)
    }

    func setSessionCaptureEnabled(_ enabled: Bool) {
        lock.lock()
        sessionCaptureEnabled = enabled
        let statusHandler = captureStatusHandler
        lock.unlock()
        statusHandler?(enabled)
    }

    func setCaptureAllowedScreens(_ screens: [String]?) {
        lock.lock()
        captureAllowedScreens = screens.map { Set($0) }
        lock.unlock()
    }

    func setCaptureBlockedScreens(_ screens: [String]) {
        lock.lock()
        captureBlockedScreens = Set(screens)
        lock.unlock()
    }

    func setScreenCaptureEnabled(_ enabled: Bool, for screen: String) {
        lock.lock()
        if enabled {
            captureBlockedScreens.remove(screen)
        } else {
            captureBlockedScreens.insert(screen)
        }
        lock.unlock()
    }

    func screen(_ name: String) {
        lock.lock()
        currentScreen = name
        lock.unlock()
        track(event: "screen_view", screen: name, elementId: nil, metadata: [:], x: nil, y: nil)
    }

    func track(
        event: String,
        screen: String? = nil,
        elementId: String?,
        metadata: [String: String],
        x: CGFloat? = nil,
        y: CGFloat? = nil
    ) {
        guard let snapshot = snapshot(screenOverride: screen) else {
            assertionFailure("Call Maze.configure(MazeConfig) before tracking events.")
            return
        }

        Task {
            let sessionId = await MazeSession.shared.currentSessionId()
            let sanitizedMetadata = metadata.mapValues { $0.count > 24 ? "***" : $0 }
            let bounds = UIScreen.main.bounds
            let normalizedX = x.map { Double(min(max($0 / bounds.width, 0), 1)) }
            let normalizedY = y.map { Double(min(max($0 / bounds.height, 0), 1)) }
            var screenshotId: String? = nil
            if event == "screen_view", snapshot.captureEnabled {
                screenshotId = await captureAndUploadScreenshot(
                    config: snapshot.config,
                    queue: snapshot.queue,
                    sessionId: sessionId,
                    screen: snapshot.screen
                )
            }

            let payload = MazeEvent(
                eventId: UUID().uuidString,
                sessionId: sessionId,
                deviceId: snapshot.config.deviceId,
                occurredAt: isoFormatter.string(from: Date()),
                event: event,
                screen: snapshot.screen,
                elementId: elementId,
                x: normalizedX,
                y: normalizedY,
                screenWidth: Double(bounds.width),
                screenHeight: Double(bounds.height),
                appVersion: snapshot.config.appVersion,
                screenshotId: screenshotId,
                metadata: sanitizedMetadata
            )
            await snapshot.queue.enqueue(payload)
        }
    }

    private func snapshot(screenOverride: String?) -> (config: MazeConfig, queue: EventQueue, screen: String?, captureEnabled: Bool)? {
        lock.lock()
        defer { lock.unlock() }
        guard let config, let queue else { return nil }
        let activeScreen = screenOverride ?? currentScreen
        return (config, queue, activeScreen, shouldCapture(screen: activeScreen))
    }

    private func shouldCapture(screen: String?) -> Bool {
        guard sessionCaptureEnabled else { return false }
        if let captureEvaluator, captureEvaluator(screen) == false { return false }
        if let screen, captureBlockedScreens.contains(screen) { return false }
        if let captureAllowedScreens, let screen {
            return captureAllowedScreens.contains(screen)
        }
        return captureAllowedScreens == nil
    }

    private func captureAndUploadScreenshot(
        config: MazeConfig,
        queue: EventQueue,
        sessionId: String,
        screen: String?
    ) async -> String? {
        let capture = await MainActor.run { () -> (data: Data, width: Int, height: Int)? in
            guard
                let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
                let window = windowScene.windows.first(where: { $0.isKeyWindow })
            else {
                return nil
            }
            let size = window.bounds.size
            guard size.width > 0, size.height > 0 else { return nil }
            let renderer = UIGraphicsImageRenderer(size: size)
            let image = renderer.image { context in
                window.layer.render(in: context.cgContext)
            }
            guard let resized = MazeClient.resize(image: image, maxDimension: config.screenshotMaxDimension),
                  let jpeg = resized.jpegData(compressionQuality: config.screenshotQuality) else {
                return nil
            }
            return (jpeg, Int(resized.size.width), Int(resized.size.height))
        }
        guard let capture else { return nil }
        do {
            return try await queue.uploadScreenshot(
                jpegData: capture.data,
                screen: screen,
                sessionId: sessionId,
                width: capture.width,
                height: capture.height
            )
        } catch {
            return nil
        }
    }

    @MainActor
    private static func resize(image: UIImage, maxDimension: CGFloat) -> UIImage? {
        let width = image.size.width
        let height = image.size.height
        let largest = max(width, height)
        if largest <= maxDimension { return image }
        let scale = maxDimension / largest
        let target = CGSize(width: width * scale, height: height * scale)
        let renderer = UIGraphicsImageRenderer(size: target)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: target))
        }
    }

    private func validateEndpoint(_ endpoint: URL) {
        let host = endpoint.host?.lowercased()
        let isLocalhost = host == "127.0.0.1" || host == "localhost"
        if endpoint.scheme?.lowercased() != "https", !isLocalhost {
            assertionFailure("Maze recommends HTTPS endpoints for production SDK traffic.")
        }
    }
}

public enum Maze {
    public static func configure(_ config: MazeConfig) {
        MazeClient.shared.configure(config)
    }

    public static func screen(_ name: String) {
        MazeClient.shared.screen(name)
    }

    public static func track(
        event: String,
        screen: String? = nil,
        elementId: String? = nil,
        metadata: [String: String] = [:],
        x: CGFloat? = nil,
        y: CGFloat? = nil
    ) {
        MazeClient.shared.track(
            event: event,
            screen: screen,
            elementId: elementId,
            metadata: metadata,
            x: x,
            y: y
        )
    }

    public static func setSessionCaptureEnabled(_ enabled: Bool) {
        MazeClient.shared.setSessionCaptureEnabled(enabled)
    }

    public static func setCaptureAllowedScreens(_ screens: [String]?) {
        MazeClient.shared.setCaptureAllowedScreens(screens)
    }

    public static func setCaptureBlockedScreens(_ screens: [String]) {
        MazeClient.shared.setCaptureBlockedScreens(screens)
    }

    public static func setScreenCaptureEnabled(_ enabled: Bool, for screen: String) {
        MazeClient.shared.setScreenCaptureEnabled(enabled, for: screen)
    }
}
